sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java

```diff
@@ -150,4 +164,104 @@ public void onFailure(Throwable e) {
+                    table.tableIdent(),
+                    repositoryInfo,
+                    ignoreUnavailable
+                ));
+            } else {
+                restoreIndices.add(table.partitionName().asIndexName());
+            }
+        }
+        return ImmutableList.copyOf(restoreIndices);
+    }
+
+    @VisibleForTesting
+    static Collection<String> indexNames(TableIdent tableIdent, RepositoryInfo repositoryInfo, boolean ignoreUnavailable) {
+        // If ignoreUnavailable is true, it's cheaper to simply return indexName and the partitioned wildcard instead
+        // checking if it's a partitioned table or not
+        String indexName = tableIdent.indexName();
+        if (ignoreUnavailable) {
+            return Arrays.asList(
+                indexName,
+                PartitionName.templateName(tableIdent.schema(), tableIdent.name()) + "*"
+            );
+        }
+        return indexNamesFromSnapshot(tableIdent, repositoryInfo.snapshots());
+    }
+
+    @VisibleForTesting
+    static Collection<String> indexNamesFromSnapshot(TableIdent tableIdent, List<SnapshotInfo> snapshots) {
+        String indexName = tableIdent.indexName();
+        for (SnapshotInfo snapshot : snapshots) {
+            for (String index : snapshot.indices()) {
+                if (indexName.equals(index)) {
+                    return Collections.singletonList(indexName);
+                } else if(isIndexPartitionOfTable(index, tableIdent)) {
+                    // add a partitions wildcard
+                    // to match all partitions if a partitioned table was meant
+                    return Collections.singletonList(
+                        PartitionName.templateName(tableIdent.schema(), tableIdent.name()) + "*");
+                }
+            }
+        }
+        throw new TableUnknownException(tableIdent);
+    }
+
+    private static boolean isIndexPartitionOfTable(String index, TableIdent tableIdent) {
+        return PartitionName.isPartition(index) &&
+               PartitionName.fromIndexOrTemplate(index).tableIdent().equals(tableIdent);
+    }
+
+    @VisibleForTesting
+    static class RepositoryInfo {
+
+        private final String repositoryName;
+        private final TransportActionProvider transportActionProvider;
+
+        List<SnapshotInfo> snapshots;
+
+        RepositoryInfo(TransportActionProvider transportActionProvider, String repositoryName) {
+            this.repositoryName = repositoryName;
+            this.transportActionProvider = transportActionProvider;
+        }
+
+        List<SnapshotInfo> snapshots() {
+            if (snapshots == null) {
+                snapshots = fetchSnapshots() ;
+            }
+            return snapshots;
+        }
+
+        private List<SnapshotInfo> fetchSnapshots() {
+            final CompletableFuture<List<SnapshotInfo>> future = new CompletableFuture<>();
+            transportActionProvider.transportGetSnapshotsAction().execute(
```

Comment:
please change the dependency to depend only on `TransportGetSnapshotsAction` instead of everything the `TransportActionProvider` class contains